CMakeLists.txt

```diff
@@ -65,22 +65,30 @@ macro(build_yaml_cpp)
   )
 endmacro()
 
-build_yaml_cpp()
+# NO_CMAKE_PACKAGE_REGISTRY used to avoid finding the library downloaded in WORKSPACE B
+# when building workspace A.
```

Comment:
What if workspace B is an underlay of workspace A? Wouldn't it be desired to use it from workspace A then?

Reply:
I don't see why both overlay and underlay would have a copy of `yaml_cpp_vendor`.
I'm not sure if that case can be handled separately.

The idea is to force cmake to look for the system installed version, and avoid the issue previously experienced on CI.

Reply:
> I don't see why both overlay and underlay would have a copy of `yaml_cpp_vendor`.

A Docker image contains a workspace building ROS 2 from source  - including `yaml_cpp_vendor`. Afterwards a user creates a workspace on top and also places `yaml_cpp_vendor` in it - either in order to get something newer or just because even though it isn't necessary.

Reply:
> A Docker image contains a workspace building ROS 2 from source - including `yaml_cpp_vendor`. Afterwards a user creates a workspace on top and also places `yaml_cpp_vendor` in it - either in order to get something newer or just because even though it isn't necessary.

I think this case is covered by `colcon\ament_cmake` doing great job to properly order `CMAKE_PREFIX_PATH`? 😉

One caveat of the package registry is that the search order is [undefined](https://gitlab.kitware.com/cmake/community/-/wikis/doc/tutorials/Package-Registry):

```
The search order among package registry entries for a single package is unspecified and the entry names (hashes in this example) have no meaning.
```

In turns, I think there is no workspace overlay concept for CMake package registry?

Reply:
> A Docker image contains a workspace building ROS 2 from source - including yaml_cpp_vendor. Afterwards a user creates a workspace on top and also places yaml_cpp_vendor in it - either in order to get something newer or just because even though it isn't necessary.

In that case, yes it will be downloaded twice.
But I don't see how I can detect that particular use case.

Using the library downloaded in another workspace can be quite confusing (if there's no underlay/overlay relation).
e.g.: build workspace A, build ws B, remove A, B now has runtime problems for finding yaml-cpp.

If you have any suggestion on how to solve that issue I'm willing to apply them, but I can't think of a way of handling that case correctly.
This PR is still an improvement over the current master, which always downloads yaml-cpp.

Reply:
> In that case, yes it will be downloaded twice.
> But I don't see how I can detect that particular use case.

A vendor package does either of the following two options:
* always build the vendored code
* check if the vendored code is already findable / usable (maybe in a new enough version) and only build it if that is not already the case

Based on that I think the vendor package in the overlay should be able to detect either a system installed version or itself in an underlay workspace. I don't see why that shouldn't be possible.

> Using the library downloaded in another workspace can be quite confusing (if there's no underlay/overlay relation).

If there is no underlay/overlay relation why would it find the library in that other workspace?

> build workspace A, build ws B, remove A, B now has runtime problems for finding yaml-cpp.

This is the case for any kind of package and has nothing to do with the CMake registry. Even when you do the same in a single workspace (build pkg C which depends on D which is also in the workspace, then delete D and try rebuilding).

Reply:
> If there is no underlay/overlay relation why would it find the library in that other workspace?

AFAIU there's only one package registry per user: https://cmake.org/cmake/help/v3.2/manual/cmake-packages.7.html#user-package-registry.

> This is the case for any kind of package and has nothing to do with the CMake registry. Even when you do the same in a single workspace (build pkg C which depends on D which is also in the workspace, then delete D and try rebuilding).

Sure, but I expect that to fail.
If `yaml_cpp_vendor` found a `yaml-cpp` library in a unrelated workspace through the cmake package registry, that would be more confusing.

Reply:
We should maybe consider a more generic approach like colcon/colcon-cmake#91 rather then a one-off patch in this repo.

Reply:
> We should maybe consider a more generic approach like [colcon/colcon-cmake#91](https://github.com/colcon/colcon-cmake/issues/91) rather then a one-off patch in this repo.

The linked example is exactly the same problem that happens here, a vendor package doing find_package quiet and then downloading/building if not found.
IMO, the problem is that the vendor package is not well implemented, not a missing feature in colcon/ament.

According to [cmake docs](https://cmake.org/cmake/help/v3.1/variable/CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY.html):

> In some cases, for example to locate only system wide installations, it is not desirable to use the User Package Registry when searching for packages

`NO_CMAKE_PACKAGE_REGISTRY` is exactly intended to solve this issue.
Even if we want to globally disable the package registry in colcon by default, the vendor package should be using this option IMO.

> I think this case is covered by colcon\ament_cmake doing great job to properly order CMAKE_PREFIX_PATH? wink

I think that's true. In the case of an overlay/underlay, the library isn't going to be downloaded twice.